tests/test_async_utils.py

```diff
@@ -0,0 +1,88 @@
+import asyncio
+
+import mock
+import pytest
+
+from paasta_tools.async_utils import async_ttl_cache
+
+
+@pytest.mark.asyncio
+async def test_async_ttl_cache_hit():
+    return_values = iter(range(10))
+
+    @async_ttl_cache(ttl=None)
+    async def range_coroutine():
+        return next(return_values)
+
+    assert await range_coroutine() == await range_coroutine()
+
+
+@pytest.mark.asyncio
+async def test_async_ttl_cache_miss():
+    return_values = iter(range(10))
+
+    @async_ttl_cache(ttl=0)
+    async def range_coroutine():
+        return next(return_values)
+
+    assert await range_coroutine() != await range_coroutine()
+
+
+@pytest.mark.asyncio
+async def test_async_ttl_cache_doesnt_cache_failures():
+    flaky_error_raiser = mock.Mock(side_effect=[Exception, None])
+
+    @async_ttl_cache(ttl=None)
+    async def flaky_coroutine():
+        return flaky_error_raiser()
+
+    with pytest.raises(Exception):
+        await flaky_coroutine()
+
+    # if we were caching failures, this would fail
+    assert await flaky_coroutine() is None
+
+
+@pytest.mark.asyncio
+async def test_async_ttl_cache_returns_in_flight_future():
+    return_values = iter(range(10))
+    condition = asyncio.Condition()
+    event = asyncio.Event()
+
+    class WaitingCoroutines:
+        count = 0
+
+    # Wait until we have enough coroutines waiting to return a result.  This
+    # ensures that dependent coroutines have a chance to get a future out of
+    # the cache
+    @async_ttl_cache(ttl=0)
+    async def range_coroutine():
+        await event.wait()
+        return next(return_values)
+
+    # Wait until we have enough coroutines waiting on range_coroutine, then
```

Comment:
I see two issues with the proposed replacement.  First, although it creates two `wait_for_event` coroutines before setting the event, those coroutines won't do anything until they are awaited, meaning the event is not really blocking anything here as I understand it (since by the time they hit `await event.wait()` the event has already been set).  Second, the coroutine returned by calling `wait_for_event()` represents a call to `async_ttl_cache`, not a call to the underlying `wait_for_event` method, so those aren't the futures we want to compare.  The futures we want to compare only exist within the `async_ttl_cache` code.

Reply:
That said, I think I am going to add a `cache` param to `async_ttl_cache` like we discussed on slack on Thursday.  This will let us dependency-inject the cache dict for testing, which would make all of this a lot simpler -- I don't have to manually create a race condition, I can just create the cache state that would have resulted from the race.